test/unittest/xtypes/XTypesTests.cpp

```diff
@@ -833,6 +840,73 @@ TEST_F(XTypesTests, MemberDescriptorFullyQualifiedName)
     EXPECT_EQ(ReturnCode_t::RETCODE_BAD_PARAMETER, my_builder->add_member(member_id++, "tJ", my_builder->build()));
 }
 
+/**
+ * This is a regression test for redmine issue #21664.
+ *
+ * It calls TypeObjectFactory::get_instance() from multiple threads to check that the factory singleton is thread safe.
+ */
+TEST(XTypesTestsThreadSafety, TypeObjectFactoryGetInstanceIsThreadSafe)
+{
+    constexpr size_t num_threads = 10;
+    std::array<TypeObjectFactory*, num_threads> factories;
+    std::array<std::thread, num_threads> threads;
+    std::atomic<size_t> n_started_threads{ 0u };
+
+    // We use a fake lock instead of a real mutex since we need the threads to act at the same time
+    struct FakeLock
+    {
+        void lock()
+        {
+        }
+
+        void unlock()
+        {
+        }
+
+    };
+    FakeLock fake_lock;
+    std::condition_variable_any cv;
+
+    // Create threads that get an instance of the factory
+    for (size_t i = 0; i < num_threads; ++i)
+    {
+        threads[i] = std::thread(
+            [&cv, &fake_lock, &n_started_threads, &factories, i]()
+            {
+                // Notify that this thread is ready
+                ++n_started_threads;
+                cv.notify_all();
```

Comment:
I think it may be clearer if we move the `cv.notify_all();` after the `for{ }`

Reply:
47adba9